api/docs_handler_test.go

```diff
@@ -0,0 +1,129 @@
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/topfreegames/pitaya-admin/api"
+	pt "github.com/topfreegames/pitaya-admin/testing"
+)
+
+func TestGetDocs(t *testing.T) {
+	t.Parallel()
+	if pt.IsConf == false {
+		pt.ConfApp()
+	}
+	a := pt.TestApp
+	var (
+		response    *httptest.ResponseRecorder
+		docsHandler *api.DocsHandler
+	)
+	docsHandler = api.NewDocsHandler(a)
+	tables := map[string]struct {
+		request *http.Request
+		asserts func(response *httptest.ResponseRecorder)
+	}{
+		"test_get_docs_no_param": {
+			request: func() *http.Request {
+				request, err := http.NewRequest("GET", "/docs", nil)
+				assert.NoError(t, err)
+				return request
+			}(),
+			asserts: func(response *httptest.ResponseRecorder) {
+				b := response.Body.String()
+				assert.Equal(t, http.StatusBadRequest, response.Code)
+				assert.Equal(t, `{"success":false, "message":"at least server type must be specified", "reason": "no server type chosen for getting autodoc"}`, b)
+			},
+		},
+		"test_get_docs_by_type_but_sv_does_not_exists": {
+			request: func() *http.Request {
+				request, err := http.NewRequest("GET", "/docs?type=bla", nil)
+				assert.NoError(t, err)
+				return request
+			}(),
+			asserts: func(response *httptest.ResponseRecorder) {
+				b := response.Body.String()
+				assert.Equal(t, http.StatusInternalServerError, response.Code)
+				assert.Equal(t, `{"success":false, "message":"failed to get docs", "reason": "no servers available of this type"}`, b)
+			},
+		},
+		"test_get_docs_success": {
+			request: func() *http.Request {
+				request, err := http.NewRequest("GET", "/docs?type=connector", nil)
+				assert.NoError(t, err)
+				return request
+			}(),
+			asserts: func(response *httptest.ResponseRecorder) {
+				b := response.Body.String()
+				assert.Equal(t, `{"handlers":{"connector.connector.getsessiondata":{"input":null,"output":[{"Data":"map[string]interface {}"},"error"]},"connector.connector.notifysessiondata":{"input":{"data":"map[string]interface {}"},"output":[]},"connector.connector.setsessiondata":{"input":{"data":"map[string]interface {}"},"output":[{"Code":"int32","Msg":"string"},"error"]}},"remotes":{"connector.connectorremote.docs":{"input":{"getProtos":"bool"},"output":[{"doc":"string"},"error"]},"connector.connectorremote.proto":{"input":{"name":"string"},"output":[{"desc":"[]byte"},"error"]},"connector.connectorremote.remotefunc":{"input":{"Msg":"string"},"output":[{"Msg":"string"},"error"]},"connector.sys.bindsession":{"input":{"data":"[]byte","id":"int64","uid":"string"},"output":[{"data":"[]byte","error":{"code":"string","metadata":"map[string]string","msg":"string"}},"error"]},"connector.sys.kick":{"input":{"userId":"string"},"output":[{"kicked":"bool"},"error"]},"connector.sys.pushsession":{"input":{"data":"[]byte","id":"int64","uid":"string"},"output":[{"data":"[]byte","error":{"code":"string","metadata":"map[string]string","msg":"string"}},"error"]}}}`, b)
```

Comment:
Won't this test be flaky? Go's maps don't have any type of order guarantees

Reply:
I think it wont be flaky since even though Go maps don't have any type of order guarantee, when marshalled as json the keys are sorted and used as JSON object keys, as stated in the (reference)[https://golang.org/pkg/encoding/json/#Marshal]

Reply:
@cscatolini Don't you think that this test will break easily when changing other minor things? Is there any real value in testing the exact response here?

Reply:
I agree with testing the response has the expected fields